tools/cli/plugin/headers_provider_plugin.go

```diff
@@ -0,0 +1,106 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"net/rpc"
+
+	"github.com/hashicorp/go-plugin"
+	"google.golang.org/grpc/metadata"
+)
+
+type HeadersProviderSDK interface {
+	GetHeaders(context.Context) (map[string]string, error)
+}
+
+type HeadersProvider interface {
+	GetHeaders(map[string][]string) (map[string]string, error)
+}
+
+type HeadersProviderRPCServer struct {
+	Impl HeadersProvider
+}
+
+type HeadersProviderPlugin struct {
+	Impl HeadersProvider
+}
+
+type HeadersProviderRPC struct {
+	client *rpc.Client
+}
+
+type HeadersProviderPluginWrapper struct {
+	provider HeadersProvider
+}
+
+func (w HeadersProviderPluginWrapper) GetHeaders(ctx context.Context) (map[string]string, error) {
+	var outgoingHeaders map[string][]string
+	if headers, ok := metadata.FromIncomingContext(ctx); ok {
```

Comment:
I think you need `FromOutgoingContext` here. Incoming context is the context which has headers from incoming request which is useful on server side only.

But actually I don't think that you need it at all. I would remove this parameter from `GetHeaders`. `AppendToOutgoingContext` will merge headers anyway. So why may one need exiting headers? I believe exiting one has `context`. @vitarb, can you clarify it?

Reply:
If you decide to keep it, `existingHeaders` would be a proper name for it.

Reply:
oops, yes, that should have been outgoing indeed. I think there will be use cases that need the headers and it would be painful to add it later so keeping it. Renamed to currentHeaders which seems to fit in with other usage in the code base.